sapicore/pipeline/__init__.py

```diff
@@ -1,21 +1,44 @@
-"""Pipeline
-===========
+"""Configurable workflows and scripts.
+
+Pipelines are flexible objects consisting of two logical parts: configurable parameters and a custom
+script defined by overriding the abstract method :meth:`~Pipeline.run`.
+
+To facilitate the development and testing of models, Sapicore provides a compact default simulation
+pipeline with a sample configuration YAML. Advanced users may use those as a basis for custom workflows.
 
-A pipeline is a class that can run the training and testing of a model.
-It should handle command line input, model configuration, and training and
-testing.
 """
+import os
+from tree_config import Configurable, load_config, apply_config
+
+__all__ = ("Pipeline",)
 
-__all__ = ('PipelineBase', )
 
+class Pipeline(Configurable):
+    """Pipeline base class.
+
+    Parameters
+    ----------
+    configuration: dict or str, optional
+        Dictionary or path to YAML to use with :meth:`apply_config`.
 
-class PipelineBase:
-    """Baseclass for pipelines.
     """
 
-    def run(self) -> None:
-        """Runs the pipeline.
+    def __init__(self, configuration: dict | str = None, **kwargs):
+        # parse configuration from YAML if given and use resulting dictionary to initialize attributes.
+        if isinstance(configuration, str) and os.path.exists(configuration):
+            apply_config(self, load_config(None, configuration))
+            self.configuration = load_config(self, filename=configuration)
 
-        To be overwritten by users.
-        """
+        elif isinstance(configuration, dict):
+            apply_config(self, configuration)
+
+        # developer may define arbitrary attributes at instantiation.
+        for key, value in kwargs.items():
+            setattr(self, key, value)
+
+    def __repr__(self):
+        return str(vars(self))
```

Comment:
While it's nice to be able to quickly see the state, printing all the variables is faaaar too much. At best we can print the config variables. But even that may be too much.

Imagine doing in the terminal:

```py
>>>print(pipeline)
```
and getting back a whole essay.

Reply:
Agreed--would you prefer to return config vars or no __repr__ at all? It's not critical for pipelines, but could be useful.

Reply:
I can see it either way, but may to start off have no `__repr__` because that's a quick fix and we can think about changes later!?

Reply:
The deferred ones, absolutely. This one will be done in this PR. I'm fine with not having `__repr__` for `Pipeline`s.